apiexport/provider.go

```diff
@@ -75,228 +71,180 @@ type Options struct {
-			if err := p.aware.Engage(clusterCtx, clusterName.String(), cl); err != nil {
-				p.log.Error(err, "failed to engage cluster", "cluster", clusterName)
 				p.lock.Lock()
-				cancel()
-				if p.clusters[clusterName] == cl {
-					delete(p.clusters, clusterName)
-					delete(p.cancelFns, clusterName)
-				}
-				p.lock.Unlock()
+				defer p.lock.Unlock()
+				p.endpoints[id] = struct{}{}
 			}
 		},
-		DeleteFunc: func(obj any) {
-			cobj, ok := obj.(client.Object)
-			if !ok {
-				tombstone, ok := obj.(toolscache.DeletedFinalStateUnknown)
-				if !ok {
-					klog.Errorf("Couldn't get object from tombstone %#v", obj)
-					return
-				}
-				cobj, ok = tombstone.Obj.(client.Object)
-				if !ok {
-					klog.Errorf("Tombstone contained object that is not expected %#v", obj)
-					return
-				}
-			}
+		UpdateFunc: func(oldObj any, newObj any) {
+			es := newObj.(*apisv1alpha1.APIExportEndpointSlice)
+			p.log.Info("updated APIExportEndpointSlice", "name", es.Name)
 
-			clusterName := logicalcluster.From(cobj)
-
-			// check if there is no object left in the index.
-			keys, err := shInf.GetIndexer().IndexKeys(kcpcache.ClusterIndexName, clusterName.String())
-			if err != nil {
-				p.log.Error(err, "failed to get index keys", "cluster", clusterName)
-				return
+			// Create a map of URLs we stored before.
+			currentUrls := map[string]bool{}
+			p.lock.RLock()
+			for id := range p.endpoints {
+				currentUrls[id] = false
 			}
-			if len(keys) == 0 {
-				p.lock.Lock()
+			p.lock.RUnlock()
 
-				// shut down individual event broadaster
-				if err := p.recorderProvider.StopBroadcaster(clusterName.String()); err != nil {
+			for _, endpoint := range es.Status.APIExportEndpoints {
+				id := HashAPIExportURL(endpoint.URL)
+				_, exists := currentUrls[id]
+
+				// Start provider if we didn't have it registered before.
+				// Else, we just mark that we found it (for cleaning up outdated providers later).
+				if !exists {
+					cfg := rest.CopyConfig(p.config)
+					cfg.Host = endpoint.URL
+
+					logger := p.log.WithValues("endpoint", id)
+					prov, err := provider.New(cfg, provider.Options{ObjectToWatch: p.object, Scheme: p.scheme, Log: &logger})
+					if err != nil {
+						p.log.Error(err, "failed to create provider")
+					}
+					if err := p.AddProvider(id, prov); err != nil {
+						p.log.Error(err, "failed to add provider")
+					}
+
+					p.log.Info("added endpoint as new provider", "endpoint", id, "url", endpoint.URL)
+
+					p.lock.Lock()
+					p.endpoints[id] = struct{}{}
 					p.lock.Unlock()
-					p.log.Error(err, "failed to stop event broadcaster", "cluster", clusterName)
-					return
+				} else {
+					currentUrls[id] = true
 				}
+			}
 
-				cancel, ok := p.cancelFns[clusterName]
-				if ok {
-					p.log.Info("disengaging cluster", "cluster", clusterName)
-					cancel()
-					delete(p.cancelFns, clusterName)
-					delete(p.clusters, clusterName)
+			// Cleanup outdated URLs.
```

Comment:
URLs that have been removed from the endpoint slice are never cleaned up in `p.endpoints`.